### ISSUE TITLE
networkd v249 interface rename conflict: Failed to manage link by its new name: File exists

### ISSUE THREAD

**Issue by frebib** [systemd/systemd#20203]
**systemd version the issue has been seen with**
 > v249 (https://salsa.debian.org/systemd-team/systemd/-/commit/a79b98cbcfe8f2e9ae088cb522add63df49789eb)

**Used distribution**
 > Debian Buster

**Linux kernel version used** (`uname -a`)

 > `Linux localhost 5.10.42-cloudflare-kasan-2021.6.4 #1 SMP Mon Sep 27 00:00:00 UTC 2010 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > amd64

**Expected behaviour you didn't see**
 > network interfaces rename their links and address configuration commences. No renaming ext0 (already renamed by udev) back to eth0.
Previous systemd build was v248.3 with all other variables the same and this didn't ever happen

**Unexpected behaviour you saw**
 > link failing to rename and moving to `failed` state.
 This only happens about ~50% of the time, and sometimes only with 1 NIC. This is the first time I've seen it happen with both interfaces. It's reproducible across several machines with varying hardware and NIC vendors. All use the same configuration.

**Steps to reproduce the problem**
 > Configure multiple .link files to match and rename the interfaces with some others left unconfigured so defaulting to eth0/eth1. In this case these are two ports on a single Solarflare NIC and the two other unconfigured interfaces are onboard and unused.
 ```
 root@localhost:/etc/systemd/network# tail -vn +1 *.link
==> 25-ext.link <==
[Match]
Driver=sfc
Path=*0

[Link]
MACAddressPolicy=persistent
Name=ext0

==> 25-int.link <==
[Match]
Driver=sfc
Path=*1

[Link]
MACAddressPolicy=persistent
Name=int0
```
**Additional program output to the terminal or log subsystem illustrating the issue**
networkd appears to be trying to rename ext0/int0 back to eth0/eth1 even though there are other unconfigured interfaces in the system already named eth0/eth1:
```
root@localhost:/etc/systemd/network# ip l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: ext0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:0f:53:49:93:a0 brd ff:ff:ff:ff:ff:ff
3: int0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:0f:53:49:93:a1 brd ff:ff:ff:ff:ff:ff
4: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether <snip> brd ff:ff:ff:ff:ff:ff
5: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether <snip> brd ff:ff:ff:ff:ff:ff
```
Previously in the kernel log, udev picks up the interface and renames in almost immediately:
```
root@localhost:/etc/systemd/network# dmesg -T | grep sfc
...
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.0 (unnamed net_device) (uninitialized): Solarflare NIC detected
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.0 (unnamed net_device) (uninitialized): no PTP support
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.1 (unnamed net_device) (uninitialized): Solarflare NIC detected
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.1 (unnamed net_device) (uninitialized): no PTP support
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.1 int0: renamed from eth1
[Mon Jul 12 09:05:46 2021] sfc 0000:86:00.0 ext0: renamed from eth0
```
then networkd starts and does some funky renames that are unexpected:
```
Jul 12 09:05:58 localhost systemd-networkd[5164]: ext0: Interface name change detected, renamed to eth0.
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: Failed to manage link by its new name: File exists
Jul 12 09:05:58 localhost systemd-networkd[5164]: Could not process link message: File exists
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: Failed
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: State changed: initialized -> failed
Jul 12 09:05:58 localhost systemd-networkd[5164]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.2 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Jul 12 09:05:58 localhost systemd-networkd[5164]: Bus bus-api-network: changing state HELLO → RUNNING
Jul 12 09:05:58 localhost systemd-networkd[5164]: int0: Interface name change detected, renamed to eth1.
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: Failed to manage link by its new name: File exists
Jul 12 09:05:58 localhost systemd-networkd[5164]: Could not process link message: File exists
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: Failed
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: State changed: initialized -> failed
Jul 12 09:05:58 localhost systemd-networkd[5164]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_33 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=4 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: Interface name change detected, renamed to int0.
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: Interface name change detected, renamed to ext0.
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: Link state is up-to-date
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth1: State changed: initialized -> unmanaged
Jul 12 09:05:58 localhost systemd-networkd[5164]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_35 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=5 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: Link state is up-to-date
Jul 12 09:05:58 localhost systemd-networkd[5164]: eth0: State changed: initialized -> unmanaged
Jul 12 09:05:58 localhost systemd-networkd[5164]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=6 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 12 09:05:58 localhost systemd-networkd[5164]: lo: Link state is up-to-date
Jul 12 09:05:58 localhost systemd-networkd[5164]: lo: State changed: initialized -> unmanaged
```

Error seems to come from here, which makes sense as `eth0` already exists: https://github.com/systemd/systemd/blob/5f9fa7a5f3eca3c1464cbc93e3508c777da14935/src/network/networkd-link.c#L2187

Full networkd debug logs: https://gist.githubusercontent.com/frebib/06cceffb1711512dc1f092438cb568f2/raw/4ca609412381d65313991d724e420bfbc24c046a

Restarting networkd or `networkctl reconfigure <interface>` seems to get things moving again, but that makes sense

**Comment by yuwata** [comment]
Hopefully, #20206 fixes this issue. If possible, please test the PR. Thank you.

**Comment by frebib** [comment]
Thanks for the quick fix! I'll give this a try first thing tomorrow.
I'm curious how this is only a problem now and not in 248.3. Would the duplicate events not have been sent beforehand?

**Comment by amishmm** [comment]
I have an issue which seems to be related to this issue.

I have reported it on systemd mailing list
First email - https://lists.freedesktop.org/archives/systemd-devel/2021-August/046764.html
Follow up email with detail - https://lists.freedesktop.org/archives/systemd-devel/2021-August/046777.html

**But I dont know if the fix of this issue is applied in systemd 249.3 (Arch Linux)**

If fix is applied that means that this fix is breaking my system.
If fix is not applied in 249.3 that means this fix may fix my issue.

How do I know this?

Thank you

**Comment by yuwata** [comment]
@amishmm The fix is already backported to v249-stable repository https://github.com/systemd/systemd-stable/commit/d2895063305712cd9e5d7f4361f9343bf3b3f00b, but 249.3 does not include it. Please wait for 249.4 or later, or backport the fix by yourself.
